Provide support for access_token refresh for flows (like OAuth2) that granted a refresh_token to a Connected App

newhouse opened the pull request:
`simple-salesforce` seems like the most vibrant of the python Salesforce "SDK" modules, but it doesn't seem to inherently support the refreshing of access tokens when a Connected App (or perhaps other flow) was provided a `refresh_token` for this very purpose. I'd written my own wrapper class around `simple-salesforce` to accomplish this, but seemed like others would have this issue/problem as well so thought I should contribute. Salesforce's documentation is all over the place, but here's a decent, succinct example that illustrates what I'm adding to the `simple-salesforce` internals here: https://help.salesforce.com/apex/HTViewHelpDoc?id=remoteaccess_oauth_refresh_token_flow.htm&language=en

I tried to add/update decent comments and update the README, but feel free to change any formatting that I screwed up or provide any feedback you'd like me to incorporate.

Perhaps adding a "access_token_refreshed" callback that can be passed in by the user and executed when a token refresh is required and happens (successfully) would be a nice feature down the line, but starting with this.

newhouse committed:
initial commit. most of the way there? (#1)

* initial commit. most of the way there?

* working. need to tidy up though

* tidied up a bit

* updating README

* formatting

* formatting

* formatting

* formatting

* formatting

Commit:
correct docs

Commit:
correct docs

Commit:
updated travis ci complaints

newhouse commented:
@nickcatal forgive me for not being super up to speed with how the CI tests work on a project like this, but I've made some changes based on the pylint, etc, but not sure if it was failing for other reasons. Can you assist? I think this branch is a valuable addition.

nickcatal commented:
I defer to @demianbrecht, who has a better sense of this, but I'm not 100% sure if we want to add so much of the OAuth flow into the library itself, or would instead prefer people to build that outside of the library and just pass in the session data that comes back from the OAuth flow.

The CI thing is weird though. I'll take a look

newhouse commented:
@nickcatal Appreciate the response and taking a look. Maybe it doesn't belong, but the solutions I've come up with that don't involve having this handled internally by ``simple-salesforce`` require some ugly and/or complicated stuff (try/excepts, class extension, complicated wrapper class/method "hacks"). My gut and investigation say to me that this is a pretty valid use case for being handled internally by a library such as this, but I could not find anything that felt right to me or pythonic anywhere out there by any library, or gist...let alone the most active and popular one here. @demianbrecht can you advise? Thanks to all.

Commit:
merged upstream master, fixed flicts

newhouse commented:
Oh sh!t: Latest merge (https://github.com/simple-salesforce/simple-salesforce/pull/124) added something called ``client_id`` to the library, but it's for a different purpose...dang Salesforce and your loosey-goosey naming conventions and consistency! Will have to re-name my stuff. No biggie. I'll be back.

newhouse committed:
change my client_ to consumer_ (#2)

* change my client_ to consumer_

* changed client_ to consumer_ in README

Commit:
pylint complaints

Commit:
more pylint

newhouse committed:
added test (#3)

Commit:
fixed mocks?

Commit:
specified content type in reposnse. fixed pylint formatting and new line

Commit:
looking for printouts

Commit:
trying tests again

newhouse commented:
OK, I've changed the names to not conflict with https://github.com/simple-salesforce/simple-salesforce/pull/124 and I've cleaned up the lint. 1/2 the tests were failing due to coverage coming at at 60 out of a min 62. So I wrote a test as well to increase the coverage, but I'm having trouble mocking out the ``requests.post`` properly as you'll see in the failed build logs. What am I doing wrong so I can get this across the finish line?

- Cheers

pitbulk commented:
I'm very interested in this feature too

I was using [salesforce-python-sdk](https://github.com/htarevern/salesforce-python-sdk), that support access_token but it seems dead and there is no UPDATE/DELETE support for RESTAPI.

I will replace the old library with simple-salesforce and newhouse I will test your code since I was using access_token.

demianbrecht commented:
My apologies for chiming in so late on this. While I'm a big +1 for adding the OAuth flows, I don't think that we really want to do it the way we auth right now, just passing in parameters. Instead, the auth mechanism should be largely redesigned (with backwards compatibility in mind as much as possible) such that each auth mechanism is its own implementation of an interface. I've jotted down some /really/ high level design thoughts in #114. I'd definitely appreciate any involvement in the actual implementation.

newhouse commented:
I can barely remember what this was really even about, and I've moved on from being in a position to need to dive back in and figure it out.

If this is still a problem, I highly encourage someone to open another PR and get 'er done. Copy my code as a start - I don't care.

It seems like this PR has become a bulletin board / placeholder for the issue instead of being the fix.

Cheers everyone

Uddipaan-Hazarika reviewed (approved):
Looks good

krishnapraveen7 commented:
Wait a minute, this is still not merged? Lmao.

fritz-astronomer commented:
fwiw, there's this https://requests-oauthlib.readthedocs.io/en/latest/oauth2_workflow.html#refreshing-tokens
and maybe something with `Salesforce(session=...)` could accomplish this? Patching that can maybe intercept the request and refresh the token as needed, transparently?

G3zz commented:
There are plenty of options to use OAuth now - as the docstrings below for the `Salesforce` class show.

```txt
        """Initialize the instance with the given parameters.
        Available kwargs
        Password Authentication:
        * username -- the Salesforce username to use for authentication
        * password -- the password for the username
        * security_token -- the security token for the username
        * domain -- The domain to using for connecting to Salesforce. Use
                    common domains, such as 'login' or 'test', or
                    Salesforce My domain. If not used, will default to
                    'login'.

        OAuth 2.0 Connected App Token Authentication:
        * consumer_key -- the consumer key generated for the user
        * consumer_secret -- the consumer secret generated for the user

        OAuth 2.0 JWT Bearer Token Authentication:
        * consumer_key -- the consumer key generated for the user

        Then either
        * privatekey_file -- the path to the private key file used
                             for signing the JWT token
        OR
        * privatekey -- the private key to use
                         for signing the JWT token

        Direct Session and Instance Access:

        * session_id -- Access token for this session

        Then either
        * instance -- Domain of your Salesforce instance, i.e.
          `na1.salesforce.com`
        OR
        * instance_url -- Full URL of your instance i.e.
          `https://na1.salesforce.com
```

I think this issue can be marked as done.
